pkg/datarepair/datarepair_test.go

```diff
@@ -271,6 +284,134 @@ func TestRepairMultipleDisqualified(t *testing.T) {
 }
 
+// TestDataRepairUploadLimits does the following:
+// - Uploads test data to nodes
+// - Get one segment of that data to check in which nodes its pieces are stored
+// - Kills as many nodes as needed which store such segment pieces
+// - Triggers data repair
+// - Verify that the number of pieces which repaired has uploaded don't overpass
+//	 the established limit (success threshold + % of excess)
+func TestDataRepairUploadLimit(t *testing.T) {
+	var repairMaxExcessRateOptimalThreshold float64
+
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount:   1,
+		StorageNodeCount: 13,
+		UplinkCount:      1,
+		Reconfigure: testplanet.Reconfigure{
+			Satellite: func(log *zap.Logger, index int, config *satellite.Config) {
+				repairMaxExcessRateOptimalThreshold = config.Repairer.MaxExcessRateOptimalThreshold
+			},
+		},
+	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+		satellite := planet.Satellites[0]
+		// stop discovery service so that we do not get a race condition when we delete nodes from overlay cache
+		satellite.Discovery.Service.Discovery.Stop()
+		satellite.Discovery.Service.Refresh.Stop()
+		// stop audit to prevent possible interactions i.e. repair timeout problems
+		satellite.Audit.Service.Loop.Stop()
+		satellite.Repair.Checker.Loop.Pause()
+		satellite.Repair.Repairer.Loop.Pause()
+
+		const (
+			repairThreshold  = 5
+			successThreshold = 7
+			maxThreshold     = 9
+		)
+		var (
+			maxRepairUploadThreshold = int(
+				math.Ceil(
+					float64(successThreshold) * (1 + repairMaxExcessRateOptimalThreshold),
+				),
+			)
+			ul       = planet.Uplinks[0]
+			testData = testrand.Bytes(8 * memory.KiB)
+		)
+
+		err := ul.UploadWithConfig(ctx, satellite, &uplink.RSConfig{
+			MinThreshold:     3,
+			RepairThreshold:  repairThreshold,
+			SuccessThreshold: successThreshold,
+			MaxThreshold:     maxThreshold,
+		}, "testbucket", "test/path", testData)
+		require.NoError(t, err)
+
+		pointer, path := getRemoteSegment(t, ctx, satellite)
+		originalPieces := pointer.GetRemote().RemotePieces
+		require.True(t, len(originalPieces) <= maxThreshold)
+
+		{ // Check that there is enough nodes in the network which don't contain
+			// pieces of the segment for being able to repair the lost pieces
+			availableNumNodes := len(planet.StorageNodes) - len(originalPieces)
+			neededNodesForRepair := maxRepairUploadThreshold - repairThreshold
+			require.Truef(t,
+				availableNumNodes >= neededNodesForRepair,
+				"Not enough remaining nodes in the network for repairing the pieces: have= %d, need= %d",
+				availableNumNodes, neededNodesForRepair,
+			)
+		}
+
+		originalStorageNodes := make(map[storj.NodeID]struct{})
+		for _, p := range originalPieces {
+			originalStorageNodes[p.NodeId] = struct{}{}
+		}
+
+		killedNodes := make(map[storj.NodeID]struct{})
+		{ // Register nodes of the network which don't have pieces for the segment
+			// to be injured and kill nodes which have pieces of such segment for
```

Comment:
instead of "kill nodes which have pieces of such segment for injuring it" it might read more clearly to say "kill nodes which have pieces of the segment in order to injure it"